components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +121,158 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
+    _TeamUtilsCommonBase<ExeSpace>(policy)
+  { }
+};
+
+/*
+ * Specialization for OpenMP execution space
+ */
 #ifdef KOKKOS_ENABLE_OPENMP
-    return omp_get_thread_num() / _team_size;
-#else
-    return 0;
-#endif
-  }
+template <>
+class TeamUtils<Kokkos::OpenMP> : public _TeamUtilsCommonBase<Kokkos::OpenMP>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy, const Real& = 1.0) :
+    _TeamUtilsCommonBase<Kokkos::OpenMP>(policy)
+  { }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& /*team_member*/) const
+  { return omp_get_thread_num() / _team_size; }
 };
+#endif
 
 /*
- * Specialization for Cuda execution space. On GPU, we expect all teams to run
- * at once.
+ * Specialization for Cuda execution space.
  */
 #ifdef KOKKOS_ENABLE_CUDA
 template <>
-class TeamUtils<Kokkos::Cuda>
+class TeamUtils<Kokkos::Cuda> : public _TeamUtilsCommonBase<Kokkos::Cuda>
 {
-  int _num_teams;
+  using Device = Kokkos::Device<Kokkos::Cuda, typename Kokkos::Cuda::memory_space>;
+  using flag_type = int; // this appears to be the smallest type that correctly handles atomic operations
+  using view_1d = typename KokkosTypes<Device>::view_1d<flag_type>;
+  using RandomGenerator = Kokkos::Random_XorShift64_Pool<Kokkos::Cuda>;
+  using rnd_type = typename RandomGenerator::generator_type;
 
-public:
+  int             _num_ws_slots;    // how many workspace slots (potentially more than the num of concurrent teams due to overprovision factor)
+  bool            _need_ws_sharing; // true if there are more teams in the policy than ws slots
+  view_1d         _open_ws_slots;    // indexed by ws-idx, true if in current use, else false
+  RandomGenerator _rand_pool;
+
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) { _num_teams = policy.league_size(); }
+  TeamUtils(const TeamPolicy& policy, const Real& overprov_factor = 1.0) :
+    _TeamUtilsCommonBase<Kokkos::Cuda>(policy),
+    _num_ws_slots(_league_size > _num_teams
+                  ? (overprov_factor * _num_teams > _league_size ? _league_size : overprov_factor * _num_teams)
+                  : _num_teams),
+    _need_ws_sharing(_league_size > _num_ws_slots),
+    _open_ws_slots("open_ws_slots", _need_ws_sharing ? _num_ws_slots : 0),
+    _rand_pool(std::chrono::high_resolution_clock::now().time_since_epoch().count())
+  { }
 
-  int get_num_concurrent_teams() const { return _num_teams; }
+  // How many ws slots are there
+  int get_num_ws_slots() const { return _num_ws_slots; }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  {
+    if (!_need_ws_sharing) {
+      return team_member.league_rank();
+    }
+    else {
+      int ws_idx = 0;
+      Kokkos::single(Kokkos::PerTeam(team_member), [&] () {
+        rnd_type rand_gen = _rand_pool.get_state(team_member.league_rank());
```

Comment:
You could save a call to get_state by executing this line only if the first atomic_c_e_s fails, since you use randomization only on the second and subsequent tries.

Reply:
Any way to avoid repeated calls to get_state in the loop without adding conditionals?

Reply:
The while condition is one conditional already. Could do something like the following and still have only one conditional:
```c++
ws_idx = team_member.league_rank() % _num_ws_slots;
for (;;) {
  if (Kokkos::atomic_compare_exchange_strong(&_open_ws_slots(ws_idx), (flag_type) 0, (flag_type)1)) {
    break;  
  } else {
    // Scope this automatic variable as narrowly as possible, in addition to calling get_state only if necessary.
    rnd_type rand_gen = _rand_pool.get_state(team_member.league_rank());
    ws_idx = Kokkos::rand<rnd_type, int>::draw(rand_gen) % _num_ws_slots;
  }
}
```

Reply:
@ambrad, i'm still a bit confused. In the implementation you suggested, you would have N-1 calls to get_state if it took N tries to find a free slot. Is that better than always 1 call to get_state?

I was thinking:
```
ws_idx = team_member.league_rank() % _num_ws_slots;
if (!Kokkos::atomic_compare_exchange_strong(&_open_ws_slots(ws_idx), (flag_type) 0, (flag_type)1)) {
  rnd_type rand_gen = _rand_pool.get_state(team_member.league_rank());
  ws_idx = Kokkos::rand<rnd_type, int>::draw(rand_gen) % _num_ws_slots;
  while (!Kokkos::atomic_compare_exchange_strong(&_open_ws_slots(ws_idx), (flag_type) 0, (flag_type)1)) {
    ws_idx = Kokkos::rand<rnd_type, int>::draw(rand_gen) % _num_ws_slots;
  }
}
```

Reply:
Oh, yes, I was confused. Yes, what you wrote gets at the original intent of my comment and also removes the error in my prototype above.